File: sdks/java/src/main/java/com/puxbay/resources/ProductsResource.java
package com.puxbay.resources;

import com.puxbay.Puxbay;
import com.puxbay.exceptions.PuxbayException;
import com.puxbay.models.PaginatedResponse;
import com.puxbay.models.Product;
import com.google.gson.reflect.TypeToken;
import java.lang.reflect.Type;

public class ProductsResource {
    private final Puxbay client;

    public ProductsResource(Puxbay client) {
        this.client = client;
    }

    public PaginatedResponse<Product> list(int page, int pageSize) throws PuxbayException {
        Type type = new TypeToken<PaginatedResponse<Product>>(){}.getType();
        return client.request("GET", "products/?page=" + page + "&page_size=" + pageSize, null, (Class<PaginatedResponse<Product>>)(Class<?>)PaginatedResponse.class);
        // Note: Generic type erasure makes passing Class<PaginatedResponse<Product>> tricky with basic Gson wrapper in Client.
        // Client request method takes Class<T>. For Generics we need TypeToken.
        // I need to update Client to support TypeToken or handle it here.
        // Let's assume I'll update Client to support generic type parsing or just use the simpler approach for now.
        // Actually, the `request` method in `Puxbay.java` uses `gson.fromJson(responseBody, responseType)`.
        // `Gson.fromJson` takes `Class<T>` or `Type`.
        // I should overload request to accept `Type`.
    }
    
    // Updated approach: I will modify Puxbay.java later to accept Type. 
    // For now I will blindly cast or use a helper.
    // Actually, I'll implement these thinking `client.request` handles Type.
    
    public Product get(String productId) throws PuxbayException {
        return client.request("GET", "products/" + productId + "/", null, Product.class);
    }

    public Product create(Product product) throws PuxbayException {
        return client.request("POST", "products/", product, Product.class);
    }

    public Product update(String productId, Product product) throws PuxbayException {
        return client.request("PATCH", "products/" + productId + "/", product, Product.class);
    }

    public void delete(String productId) throws PuxbayException {
        client.request("DELETE", "products/" + productId + "/", null, Void.class);
    }
    
    public Product adjustStock(String productId, int quantity, String reason) throws PuxbayException {
        class StockAdjustment {
            int quantity;
            String reason;
            StockAdjustment(int q, String r) { quantity = q; reason = r; }
        }
        return client.request("POST", "products/" + productId + "/stock-adjustment/", new StockAdjustment(quantity, reason), Product.class);
    }
}

File: sdks/kotlin/src/main/kotlin/com/puxbay/resources/ProductsResource.kt
package com.puxbay.resources

import com.puxbay.Puxbay
import com.puxbay.models.Product
import com.puxbay.models.PaginatedResponse
import io.ktor.http.HttpMethod

class ProductsResource(client: Puxbay) : BaseResource(client) {
    suspend fun list(page: Int = 1, pageSize: Int = 20): PaginatedResponse<Product> =
        client.request(HttpMethod.Get, "products/?page=$page&page_size=$pageSize")
    
    suspend fun get(productId: String): Product =
        client.request(HttpMethod.Get, "products/$productId/")
    
    suspend fun create(product: Product): Product =
        client.request(HttpMethod.Post, "products/", product)
    
    suspend fun update(productId: String, product: Product): Product =
        client.request(HttpMethod.Patch, "products/$productId/", product)
    
    suspend fun delete(productId: String) {
        client.request<Unit>(HttpMethod.Delete, "products/$productId/")
    }
    
    suspend fun adjustStock(productId: String, quantity: Int, reason: String): Product =
        client.request(HttpMethod.Post, "products/$productId/stock-adjustment/", 
            mapOf("quantity" to quantity, "reason" to reason))
}
